pennylane/ops/op_math/prod.py

```diff
@@ -206,8 +207,22 @@ def eigvals(self):
 
         return qml.math.prod(eigvals, axis=0)
 
-    def matrix(self, wire_order=None):
+    def matrix(self, wire_order=None, cache=False):
         """Representation of the operator as a matrix in the computational basis."""
+        wire_order = wire_order or self.wires
+        mat_hash = hash(Wires(wire_order))
+
+        if mat_hash in self._mat_cache:
+            return self._mat_cache[mat_hash]
+
+        if hash(self.wires) in self._mat_cache:
+            res = math.expand_matrix(
+                self._mat_cache[hash(self.wires)], self.wires, wire_order=wire_order
+            )
+            if cache:
+                self._mat_cache[hash(Wires(wire_order))] = res
+            return res
+
```

Comment:
I would move this all to a method such as `get_cached_matrix(self, wire_order, cache)` in composite.py. Then Prod and Sum's `matrix()` implementations will call this member function before moving on. Furthermore, will it be common that there is something to read from the cache if the `cache` param is False? It feels a little strange that we read when it's either True or False, but only write when it's False. if reads and writes are the same, the code logic could be cleaner:
- composite adds new abstract member function `compute_matrix` (the computation part below)
- composite has `get_cached_matrix`. if cache misses here, it calls `self.compute_matrix()` which the child class has implemented, caches and returns it
- `child.matrix()` says `if cache: return self.get_cached_matrix(); else: return self.compute_matrix()`

Reply:
Let me know if the current implementation matches this

Reply:
perfect, love it! making `compute_matrix` a staticmethod is causing for a bit of messiness but I'm alright with it. idk how worried we are with inlining (performance-wise) but `composite.matrix()` is just `composite._get_cached_matrix()` now that you've realized my proposal so cleanly so it could be inlined.